gax/src/main/java/com/google/api/gax/rpc/BidiStreamingCallable.java

```diff
@@ -44,22 +44,196 @@
 
+  /**
+   * The "base" method from which other forms of {@code call}s are derived. Most users will not need
+   * to call this method directly.
+   *
+   * <p>However, it is {@code public}, since library authors might want to call this method in
+   * adaptor classes.
+   */
+  public abstract ClientStream<RequestT> internalCall(
+      ResponseObserver<ResponseT> responseObserver,
+      ClientStreamReadyObserver<RequestT> onReady,
+      ApiCallContext context);
+
+  /**
+   * Listens to server responses and send requests when the network is free. Example usage:
+   *
+   * <pre>{@code
+   * final Iterator<Integer> sourceDataIterator = intCollection.iterator();
+   * BidiStreamObserver<Integer, String> bidiStreamObserver = new BidiStreamObserver<Integer, String>() {
+   *   public void onStart(StreamController controller) {
+   *     // no-op
+   *   }
+   *
+   *   public void onResponse(String response) {
+   *     System.out.println(response);
+   *   }
+   *
+   *   public void onComplete() {
+   *     System.out.println("done!");
+   *   }
+   *
+   *   public void onError(Throwable t) {
+   *     System.out.println("error: " + t);
+   *   }
+   *
+   *   public void onReady(ClientStream<Integer> stream) {
+   *     while (sourceDataIterator.hasNext()) {
+   *       if (stream.isReady()) {
+   *         stream.send(sourceDataIterator.next());
+   *       } else {
+   *         // It's OK we haven't consumed the whole iterator;
+   *         // onReady will be called again when the network becomes free.
+   *         return;
+   *       }
+   *     }
+   *     // We ran out of things to send.
+   *     stream.close();
+   *   }
+   * };
+   *
+   * bidiStreamingCallable.call(bidiStreamObserver);
+   * }</pre>
+   */
+  public void bidiCall(final BidiStreamObserver<RequestT, ResponseT> bidiObserver) {
+    bidiCall(bidiObserver, null);
+  }
+
+  /** Listens to server responses and send requests when the network is free. */
+  public void bidiCall(
+      final BidiStreamObserver<RequestT, ResponseT> bidiObserver, ApiCallContext context) {
+    internalCall(
+        bidiObserver,
+        new ClientStreamReadyObserver<RequestT>() {
+          @Override
+          public void onReady(ClientStream<RequestT> stream) {
+            bidiObserver.onReady(stream);
+          }
+        },
+        context);
+  }
+
+  /**
+   * Send requests and iterate over server responses.
+   *
+   * <p>This returns a live stream that must either be fully consumed or cancelled. Example usage:
+   *
+   * <pre>{@code
+   * BidiStream<String, String> stream = bidiStreamingCallable.call()
+   * for (String s : stream) {
+   *   if ("needle".equals(s)) {
+   *     // Cancelling the stream will cause `hasNext()` to return false on the next iteration,
+   *     // naturally breaking the loop.
+   *     stream.cancel();
+   *   }
+   *   stream.send(s);
+   * }
+   * }</pre>
+   */
+  public BidiStream<RequestT, ResponseT> call() {
```

Comment:
Looking at all the functions together, it feels arbitrary that we have one named `bidiCall` and the rest named `call`. How about we name all of these `bidiCall`? I'm comfortable with the methods in `ServerStreamingCallable` retaining the method name `call`, and in the future, also using simply `call` in `ClientStreamingCallable`.

Reply:
If we call everything bidi call, we have the same issue as discussed in [here](https://github.com/googleapis/gax-java/pull/540#discussion_r192869079) right?

Reply:
Damn you're right.